src/library/scala/collection/immutable/NumericRange.scala

```diff
@@ -132,6 +132,60 @@ sealed class NumericRange[T](
   // are forgiving: therefore the checks are with the methods.
   private def locationAfterN(n: Int): T = start + (step * fromInt(n))
 
+  private def crossesTheEndAfterN(n: Int): Boolean = {
+    // if we're sure that subtraction in the context of T won't overflow, we use this function
+    // to calculate the length of the range
+    def unsafeRangeLength(r: NumericRange[T]): T = {
+      val diff = num.minus(r.end, r.start)
+      val quotient = num.quot(diff, r.step)
+      val remainder = num.rem(diff, r.step)
+      if (!isInclusive && num.equiv(remainder, num.zero)) quotient else num.plus(quotient, num.one)
+    }
+    val stepIsInTheSameDirectionAsStartToEndVector =
+      (num.gt(end, start) && num.sign(step) == num.one) || (num.lt(end, start) && num.sign(step) == -num.one)
+
+    if (num.equiv(start, end) || n <= 0 || !stepIsInTheSameDirectionAsStartToEndVector) return n >= 1
+
+    val smallerThanInt = num match {
+      case _: Numeric.ShortIsIntegral | _: Numeric.ByteIsIntegral => true
+      case _                                                      => false
+    }
+
+    val sameSign = num.equiv(num.sign(start), num.sign(end))
+
+    if (sameSign) { // subtraction is safe
+      val len = unsafeRangeLength(this)
+
+      if (smallerThanInt) n >= num.toInt(len) else num.gteq(num.fromInt(n), len)
+    } else {
+      // split to two ranges, which subtraction is safe in both of them (around zero)
+      val closestToZeroFromStart = num.rem(start, step)
+
+      val (l: NumericRange[T], r: NumericRange[T]) = if (num.lt(start, num.zero)) {
+        val lastElementBeforeZero =
+          if (num.equiv(closestToZeroFromStart, num.zero)) step else closestToZeroFromStart
+
+        (NumericRange.inclusive(start, zero - num.abs(lastElementBeforeZero), step), copy(zero, end, step))
+      } else {
+        val firstElementAfterZero =
+          if (num.equiv(closestToZeroFromStart, num.zero)) num.abs(step) else num.abs(closestToZeroFromStart)
+
+        (copy(zero, end, step), NumericRange.inclusive(start, firstElementAfterZero, step))
+      }
+
+      val leftLength = unsafeRangeLength(l)
+      val rightLength = unsafeRangeLength(r)
+
+      if (smallerThanInt) {
+        n >= num.toInt(leftLength) + num.toInt(rightLength)
+      } else {
+        // instead of n >= rightLength + leftLength which may cause addition overflow,
+        // this can be used (n - rightLength) >= leftLength
```

Comment:
The right/left on the comment doesn't match the right/left in the code.  Either way works, but better not to say one and do the other!